crosshair/libimpl/builtinslib.py

```diff
@@ -1937,6 +1937,29 @@ def find(self, substr, start=None, end=None):
         start = 0 if start is None else force_to_smt_sort(start, SmtInt)
         return SmtInt(z3.IndexOf(smt_mystr, smt_substr, start))
 
+    def rfind(self, substr, start=None, end=None):
+        space = self.statespace
+        sub = force_to_smt_sort(substr, SmtStr)
+
+        start = 0 if start is None else force_to_smt_sort(start, SmtInt)
+        end = z3.Length(self.var) if end is None else force_to_smt_sort(end, SmtInt)
+        value = z3.SubString(self.var, start, end - 1)
+        if space.smt_fork(z3.Contains(value, sub)):
+            match_index = z3.Int(f"match_index_{space.uniq()}")
+            last_match = z3.SubString(value, match_index, z3.Length(sub))
+            index_remaining = match_index + z3.Length(sub)
+            remaining = z3.SubString(
+                value, index_remaining, z3.Length(value) - index_remaining
+            )
+            space.add(
+                z3.And(
+                    z3.Contains(last_match, sub), z3.Not(z3.Contains(remaining, sub))
```

Comment:
1. RE `z3.Contains(last_match, sub)`: Even more than containment? Are `last_match` and `sub` guaranteed to be the same string at this point?
2. RE `3.Not(z3.Contains(remaining, sub))`: help me think about whether matches to the right could overlap with the matched portion of the string. e.g. do we prevent `"aaa".rfind("aa")` from returning zero?

Reply:
1. Yes, because `last_match` is defined by taking a substring of length `sub`, so it suffices to check that `last_match` contains `sub`. I really should take a good look at the code to get it in a bit more readable order.
2. Good catch! Redefining `index_remaining = match_index+1` will take care of this problem in general, I think.